feat: initial support for Temporal equality

dirkluijk opened the pull request:
### Description

Fixes https://github.com/vitest-dev/vitest/issues/7991

<!-- You can also add additional context here -->

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [X] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [X] Ideally, include a test that fails without this PR but passes with it.
- [X] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [X] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [X] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

hi-ogawa commented on test/core/test/expect.test.ts:
```diff
@@ -346,3 +346,42 @@ describe('iterator', () => {
   })
 })
+
+describe('Temporal equality', () => {
```
Thanks for PR! I think we should test against the polyfill which is supposed to be spec compliant.

hi-ogawa commented on packages/expect/src/jest-utils.ts:
```diff
@@ -159,6 +159,15 @@ function eq(
+    case '[object Temporal.PlainYearMonth]':
+    case '[object Temporal.PlainMonthDay]':
+    case '[object Temporal.Duration]':
+      return a.toString() === b.toString()
```
Do you know which is preferred whether `toString` equality and builtin `equals` method? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal/ZonedDateTime/equals

It looks like there are several old issues mentions calendar comparison (such as https://github.com/tc39/proposal-temporal/issues/625), but the discussion seem quite technical :sweat_smile:

dirkluijk replied:
`.equals()` seems to compare including the calendar, so I think it's a good suggestion. Let's do it! 

For `Temporal.Duration` this `.equals()` method doesn't exist, but I wasn't sure about that one to begin with. For Durations, there IS actually ambiguity. For example, are 60 seconds and 1 minute the same? If you compare the duration in time, yes; but if you literally compare all their components, no.

My suggestion is to leave out Duration for now.

hi-ogawa replied:
Indeed it looks like Duration is tricky. https://github.com/tc39/proposal-temporal/issues/856 seems to explain the details.

As a builtin equality, I think choosing `toString` comparison makes sense since that's the most strict way to do it. Users can still add own custom equality tester to loosen it if desired.

dirkluijk commented:
@hi-ogawa I updated the PR:

* added Temporal polyfill
* adjusted equality check and use `.equals()`
* for `Duration` I still compare the string representation to compare all its components.

hi-ogawa reviewed (approved):
Thanks!

justingrant commented:
I'm one of the champions of the Temporal proposal. Sorry I was too late to comment on this PR back in May, but I was very happy to see this PR!  Thanks for building it. 

I did have a question for @hi-ogawa and @dirkluijk: in vitest, is equality comparison expected to have no observable differences between the values?

If yes, then using `Temporal.ZonedDateTime.prototype.equals` might be a problem. Two `ZonedDateTime` instances can use time zone IDs that are aliases of each other in the IANA Time Zone Database, for example `Asia/Calcutta` and `Asia/Kolkata`. The `equals` method will canonicalize the time zone ID before comparing, so those two time zones (and hence the ZDT instances) will be considered equal. But the value of the `timeZoneId` property on the ZDT instances will be different.

That said, this may be desirable behavior. One of the main reasons that we made `equals` canonicalize time zones is to avoid breaking automated tests when a time zone ID is deprecated in favor of a new name, like the recent rename of Europe/Kiev => Europe/Kyiv.  If you use `equals`, then you're guaranteed that even if the IANA Time Zone Database changes, your tests will still behave the same.

So the behavior in this PR may be better!  But at least I wanted to flag the behavior in case you didn't want it.

hi-ogawa commented:
> is equality comparison expected to have no observable differences between the values?

@justingrant Thanks for the heads up! For this case, Vitest's goal is to provide whatever most ergonomic for users and we don't necessary need to detect anything observable. Since Temporal provides one "opinionated" equality implementation, I think test/assertion framework building up on it sounds most natural. For any niche use case, Vitest allows overriding equality via `addEqualityTesters`, which can always serve as escape hatch.
